src/cargo/sources/path.rs

```diff
@@ -392,38 +393,66 @@ impl<'cfg> PathSource<'cfg> {
         filter: &mut dyn FnMut(&Path, bool) -> CargoResult<bool>,
     ) -> CargoResult<()> {
-        let is_dir = path.is_dir();
-        if !is_root && !(*filter)(path, is_dir)? {
-            return Ok(());
-        }
-        if !is_dir {
-            ret.push(path.to_path_buf());
-            return Ok(());
-        }
-        // Don't recurse into any sub-packages that we have.
-        if !is_root && path.join("Cargo.toml").exists() {
-            return Ok(());
+        fn walk(
+            path: &Path,
+            ret: &mut Vec<PathBuf>,
+            ancestors: &mut Vec<PathBuf>,
+            is_root: bool,
+            filter: &mut dyn FnMut(&Path, bool) -> CargoResult<bool>,
+        ) -> CargoResult<()> {
+            let is_dir = path.is_dir();
+            if !is_root && !(*filter)(path, is_dir)? {
+                return Ok(());
+            }
+            if !is_dir {
+                ret.push(path.to_path_buf());
+                return Ok(());
+            }
+
+            detect_loop(ancestors, path).with_context(|| format!("cannot read {:?}", path))?;
+
+            // Don't recurse into any sub-packages that we have.
+            if !is_root && path.join("Cargo.toml").exists() {
+                return Ok(());
+            }
+
+            let entries = fs::read_dir(path)
+                .with_context(|| format!("cannot read {:?}", path))?
+                .map(|e| e.unwrap().path());
+
+            ancestors.push(path.to_path_buf());
+            for path in entries {
+                let name = path.file_name().and_then(|s| s.to_str());
+                if is_root && name == Some("target") {
+                    // Skip Cargo artifacts.
+                    continue;
+                }
+                walk(&path, ret, ancestors, false, filter)?;
+            }
+            ancestors.pop();
+
+            Ok(())
         }
 
-        // For package integration tests, we need to sort the paths in a deterministic order to
-        // be able to match stdout warnings in the same order.
-        //
-        // TODO: drop `collect` and sort after transition period and dropping warning tests.
-        // See rust-lang/cargo#4268 and rust-lang/cargo#4270.
-        let mut entries: Vec<PathBuf> = fs::read_dir(path)
-            .with_context(|| format!("cannot read {:?}", path))?
-            .map(|e| e.unwrap().path())
-            .collect();
-        entries.sort_unstable_by(|a, b| a.as_os_str().cmp(b.as_os_str()));
-        for path in entries {
-            let name = path.file_name().and_then(|s| s.to_str());
-            if is_root && name == Some("target") {
-                // Skip Cargo artifacts.
-                continue;
+        fn detect_loop(ancestors: &[PathBuf], path: &Path) -> io::Result<()> {
+            use same_file::Handle;
+            let handle = Handle::from_path(path)?;
+            for ancestor in ancestors.iter().rev() {
+                if Handle::from_path(ancestor)? == handle {
```

Comment:
This looks like it's recreating the `Handle` on each recursive depth, but could the ancestors list instead already be a list (or set) of `Handle` values?